src/routes/jira/workspaces/repositories/jira-workspaces-repositories-get.ts

```diff
@@ -1,3 +1,108 @@
-export const JiraWorkspacesRepositoriesGet = () => {
-	// Content included in ARC-2116-search-connected-repos
+import { Request, Response } from "express";
+import { Errors } from "config/errors";
+import { Subscription } from "models/subscription";
+import { RepoSyncState } from "models/reposyncstate";
+import { transformRepositoryId } from "~/src/transforms/transform-repository-id";
+
+const {
+	MISSING_JIRA_HOST,
+	MISSING_SUBSCRIPTION,
+	MISSING_REPO_NAME,
+	NO_MATCHING_REPOSITORIES
+} = Errors;
+
+export interface WorkspaceRepo {
+	id: string,
+	name: string,
+	workspaceId: string
+}
+
+const findMatchingRepos = async (
+	jiraHost: string,
+	repoName: string,
+	connectedOrgId: number | undefined
+) => {
+	let repos;
+
+	if (connectedOrgId) {
+		const subscription = await Subscription.getOneForSubscriptionIdAndHost(jiraHost, connectedOrgId);
+
+		if (!subscription) {
+			return null;
+		}
+
+		repos = await RepoSyncState.findRepositoriesBySubscriptionIdAndRepoName(subscription.id, repoName);
+	} else {
+		const subscriptions = await Subscription.getAllForHost(jiraHost);
+
+		if (!subscriptions.length) {
+			return null;
+		}
+
+		const reposArray = await Promise.all(subscriptions.map(async (subscription) => {
+			return await RepoSyncState.findRepositoriesBySubscriptionIdAndRepoName(subscription.id, repoName);
+		}));
+
+		repos = reposArray.flat();
+	}
+
+	return repos;
+};
+
+const paginatedRepositories = (page: number, limit: number, repositories: WorkspaceRepo[]) => {
+	const startIndex = (page - 1) * limit;
+	const endIndex = page * limit;
+	return repositories.slice(startIndex, endIndex);
+};
+
+export const JiraWorkspacesRepositoriesGet = async (req: Request, res: Response): Promise<void> => {
+	req.log.info({ method: req.method, requestUrl: req.originalUrl }, "Request started to GET repositories");
+
+	const { jiraHost } = res.locals;
+
+	if (!jiraHost) {
+		req.log.warn({ jiraHost, req, res }, MISSING_JIRA_HOST);
+		res.status(400).send(MISSING_JIRA_HOST);
+		return;
+	}
+
+	const connectedOrgId = Number(req.query?.workspaceId);
+	const repoName = req.query?.searchQuery as string;
+	const page = Number(req.query?.page) || 1; // Current page (default: 1)
+	const limit = Number(req.query?.limit) || 20; // Number of items per page (default: 20)
+
+	if (!repoName) {
+		req.log.warn(MISSING_REPO_NAME);
+		res.status(400).send(MISSING_REPO_NAME);
+		return;
+	}
+
+	const repos = await findMatchingRepos(jiraHost, repoName, connectedOrgId);
+
+	if (repos === null) {
+		req.log.warn(MISSING_SUBSCRIPTION);
+		res.status(400).send(MISSING_SUBSCRIPTION);
+		return;
+	}
+
+	if (!repos?.length) {
+		req.log.warn(NO_MATCHING_REPOSITORIES);
+		res.status(400).send(NO_MATCHING_REPOSITORIES);
+		return;
+	}
+
+	const repositories: WorkspaceRepo[] = repos.map((repo) => {
+		const repoId = transformRepositoryId(repo.repoId);
```

Comment:
Currently using `transformRepositoryId` but it's effectively just returning the repositoryId when base url is passed as in this case. For most customers this won't be an issue but for any customers who have connected both cloud and server / multiple servers we could potentially run into an issue with non-unique repo ids.

We can't return to id from RepoSyncState, or prefix this onto the repoId, as this value changes anytime a customer runs a backfill. We could prefix with the subscriptionId as this value remains unchanged during backfill and if a user connects and disconnects. It will only change if a user removes permissions for the org in GH or uninstalls & reinstalls the app but think this may be our best option. Thoughts?

Reply:
Actually, how about the uniqueness of the workspace? Is it by id or name(str)? 
As for GHES, the orgName can be duplicate from different GHES server, even the workspaceId is not unique.

We have a similar thing when sending data to jira. We then use `TransformedRepositoryId` inside, `return ${calculatePrefix(gitHubBaseUrl)}-${repositoryId} as TransformedRepositoryId;`

I suspect we can do something like this? 

But might need to ask <insert_that_proper_team_name_here> for suggestions.

Reply:
Updated for this. Will need to chat to the incredibles about how they are going to handle the changing workspaceIds